lib/spack/spack/test/concretize.py

```diff
@@ -22,190 +22,184 @@
-            self.assertTrue(name in concrete.variants)
 
-        for flag in concrete.compiler_flags.valid_compiler_flags():
-            self.assertTrue(flag in concrete.compiler_flags)
 
-        if abstract.compiler and abstract.compiler.concrete:
-            self.assertEqual(abstract.compiler, concrete.compiler)
-
-        if abstract.architecture and abstract.architecture.concrete:
-            self.assertEqual(abstract.architecture, concrete.architecture)
-
-    def check_concretize(self, abstract_spec):
-        abstract = Spec(abstract_spec)
-        concrete = abstract.concretized()
-
-        self.assertFalse(abstract.concrete)
-        self.assertTrue(concrete.concrete)
-        self.check_spec(abstract, concrete)
-
-        return concrete
-
-    def test_concretize_no_deps(self):
-        self.check_concretize('libelf')
-        self.check_concretize('libelf@0.8.13')
-
-    def test_concretize_dag(self):
-        self.check_concretize('callpath')
-        self.check_concretize('mpileaks')
-        self.check_concretize('libelf')
+def check_spec(abstract, concrete):
+    if abstract.versions.concrete:
+        assert abstract.versions == concrete.versions
+
+    if abstract.variants:
+        for name in abstract.variants:
+            avariant = abstract.variants[name]
+            cvariant = concrete.variants[name]
+            assert avariant.value == cvariant.value
+
+    if abstract.compiler_flags:
+        for flag in abstract.compiler_flags:
+            aflag = abstract.compiler_flags[flag]
+            cflag = concrete.compiler_flags[flag]
+            assert set(aflag) <= set(cflag)
+
+    for name in abstract.package.variants:
+        assert name in concrete.variants
+
+    for flag in concrete.compiler_flags.valid_compiler_flags():
+        assert flag in concrete.compiler_flags
+
+    if abstract.compiler and abstract.compiler.concrete:
+        assert abstract.compiler == concrete.compiler
+
+    if abstract.architecture and abstract.architecture.concrete:
+        assert abstract.architecture == concrete.architecture
+
+
+def check_concretize(abstract_spec):
+    abstract = Spec(abstract_spec)
+    concrete = abstract.concretized()
+    assert not abstract.concrete
+    assert concrete.concrete
+    check_spec(abstract, concrete)
+    return concrete
+
+
+@pytest.fixture(
+    params=[
+        # no_deps
+        'libelf', 'libelf@0.8.13',
+        # dag
+        'callpath', 'mpileaks', 'libelf',
+        # variant
+        'mpich+debug', 'mpich~debug', 'mpich debug=2', 'mpich',
+        # compiler flags
+        'mpich cppflags="-O3"',
+        # with virtual
+        'mpileaks ^mpi', 'mpileaks ^mpi@:1.1', 'mpileaks ^mpi@2:',
+        'mpileaks ^mpi@2.1', 'mpileaks ^mpi@2.2', 'mpileaks ^mpi@2.2',
+        'mpileaks ^mpi@:1', 'mpileaks ^mpi@1.2:2'
+    ]
+)
+def spec(request):
+    """Spec to be concretized"""
+    return request.param
+
+
+@pytest.mark.usefixtures('config', 'builtin_mock')
```

Comment:
> That is good, although I am more concerned with other potential issues with having a generically-named variable like "config" available in the global namespace.

I agree if we want to change the name of fixtures for something better. For 'config' if there's general agreement over something else I'll make the change (didn't do it now as it touches a lot of files and would like to have everybody agreeing on that).

Just to be sure this is clear to everybody reading the thread: the fixture `config` *is not available in the global namespace*. `pytest` has a dependency injection mechanism (commonly referred to as *funcargs*) which works more or less like this: 

1. the decorator `@pytest.fixture` registers a function as a fixture (this function will be used as a factory to create a fixture object when required)
2. since tests are *automatically run* by `pytest` they can only accept fixture arguments: if the name of an argument matches the name of a registered fixture, an object will be *injected*
3. fixtures can be overriden with [these rules](http://doc.pytest.org/en/latest/fixture.html#overriding-fixtures-on-various-levels)